docs/faas/aws-lambda.md

```diff
@@ -66,6 +66,10 @@ contain an incomplete trace context which indicates X-Ray isn’t enabled. The e
 `Context` will be valid and sampled only if AWS X-Ray has been enabled for the Lambda function. A user can
 disable AWS X-Ray for the function if the X-Ray Span Link is not desired.
 
+Instrumentation SHOULD include an explicit option to use `Context` in `_X_AMZN_TRACE_ID` as parent insted of `Link`,
```

Comment:
```suggestion
Instrumentation SHOULD include an explicit option to use `Context` in `_X_AMZN_TRACE_ID` as parent instead of `Link`,
```

Reply:
It's my understanding this is a temporary stop-gap to help existing user migrate.

We should use similar language/terminology as was done in http semconv about migration period here.  See: https://github.com/open-telemetry/semantic-conventions/tree/main/docs/http

Reply:
Really? I think having an option to use the X-Ray context as parent seems OK also long-term. Though I would change the behavior slightly to always include the link in addition.

Reply:
More than just "OK", it is imperative that X-Ray users have the ability to utilize the X-Ray span context that is provided to them by the Lambda service.  We cannot have a system that makes that impossible and should not have one that leaves it to the instrumentation author to decide whether that category of users should be supported.  

What benefit would be gained from having the same span context both as parent and as a link?

Reply:
I don't see that as a significant benefit in exchange for the duplication of data in the span object.  Duplicating data here leads to increased costs for the user who has to move the duplicated data around.  This has been a recurring pain point for users.

I think it would be relevant to discuss this in a general spec setting if we want to have information on the span indicating which context source was used to obtain the parent context as there are many ways that a span could have obtained a parent context from multiple different sources.

Reply:
I think I agree w/ @Aneurysm9 here. As I dive into the issue more I think we're talking about *default* parenting and default scenarios.

X-ray causes problems with tracing journeys that span x-ray boundaries of trace_id synthesis, if I'm understanding correctly.  SO, users need the ability to either limit traces to x-ray's boundaries *or* allow other propagation decisions to overrule x-ray's.

I think Carlos's PR here looks like the best path forward for flexibility and targeting general usage patterns.

Reply:
> X-ray causes problems with tracing journeys that span x-ray boundaries of trace_id synthesis, if I'm understanding correctly. SO, users need the ability to either limit traces to x-ray's boundaries _or_ allow other propagation decisions to overrule x-ray's.

I'm not sure this is exactly correct.  Yes, AWS X-Ray does have requirements for trace ID values that are not shared by other backends.  That is orthogonal to the X-Ray context propagation format, which can be used to encode and propagate contexts including trace ID values that do not conform to the requirements of the AWS X-Ray service.

The issue here is, in my view, one of how the instrumentation determines a carrier to be used to extract a context for the initial span created by the instrumentation's function wrapper.  That's the problem that I raised in [the initial issue](https://github.com/open-telemetry/opentelemetry-specification/issues/3060) and that I thought was resolved thusly:

> *    Instrumentation should be responsible for capturing the environment variable and putting it into the carrier before passing along to the propagator. (For implementations that have `EventToCarrier` it can be done there.)
> *    If both the environment variable and the event have x-ray propagation, then the priority between those is currently undefined and needs to be discussed further.

The [spec PR](https://github.com/open-telemetry/opentelemetry-specification/pull/3166) that followed did not adhere to that understanding and removed any mention of context carriers after I had approved it.

Following discussion of this issue at the FaaS SIG meeting last week we agreed to try to modify the existing specification language minimally in order to achieve a result that is likely to make nobody entirely satisfied but everyone could live with.  More substantive changes can be entertained, but I think that we need to agree to introduce no further breaking changes to the existing instrumentation implementations until such time as everyone involved agrees with the path forward.

Reply:
> I think Carlos's PR here looks like the best path forward for flexibility and targeting general usage patterns.

This PR is, in my opinion, too flexible as it has [no requirements, only recommendations](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1411#issuecomment-1671916117).  One of the points of contention we have had in discussions on implementation PRs is the assertion by some that this behavior is required by the spec and by others that the spec has no requirements.  https://github.com/open-telemetry/semantic-conventions/pull/272 attempts to remedy this by imposing actual requirements on instrumentation.  The default behavior remains as a recommendation.

Reply:
> that I thought was resolved thusly:
>> * Instrumentation should be responsible for capturing the environment variable and putting it into the carrier before passing along to the propagator. (For implementations that have EventToCarrier it can be done there.)
>> * If both the environment variable and the event have x-ray propagation, then the priority between those is currently undefined and needs to be discussed further.

This doesn't address at all my concern (also raised on the issue):

If *ALL OF* the follow are true, what happens:

- Environment variable has x-ray propagation
- Event has x-ray propagation
- Event has *another propgataion format supported by otel*


Users need a way to configure which one of these to prioritize (and ideally use links to represent the others).  Your proposal de-facto promote x-ray propagation while turning down native otel propagation formats.  I think we should always prefer native otel propagation formats (like w3c traceparent) in lieu of user configuration.

This is also the approach we recommend for Google Cloud Run users, where google has a custom propagation format.

Reply:
> I think that we need to agree to introduce no further breaking changes to the existing instrumentation implementations until such time as everyone involved agrees with the path forward

+1

> This PR is, in my opinion, too flexible as it has [no requirements, only recommendations](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1411#issuecomment-1671916117). One of the points of contention we have had in discussions on implementation PRs is the assertion by some that this behavior is required by the spec and by others that the spec has no requirements. #272 attempts to remedy this by imposing actual requirements on instrumentation. The default behavior remains as a recommendation.

I think some of the flexibility across the semantic conventions is intentional in order to cover a broader range of (future) third-party instrumentations. I personally expect all instrumentations published under the OpenTelemetry org to follow all "Recommendeds" and SHOULDs, though I don't think we've had this discussion as a community (I've added this question to next week's semconv meeting).

Reply:
> If ALL OF the follow are true, what happens:
>
> *    Environment variable has x-ray propagation
> *    Event has x-ray propagation
> *    Event has *another propgataion format supported by otel*

In this situation I would expect the user's configured propagator to extract a context it can understand from the carrier provided to it.  I would expect the instrumentation to construct that carrier from the event in such a way that an X-Ray context carried in the event would take precedence over the context carried in the environment.

I think this is one of the situations the [`Composite Propagator`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#composite-propagator) is specified to address.  To the extent that instrumentation allows the user to specify a composite propagator then the user is fully in control of the precedence used by that propagator.

> Your proposal de-facto promote x-ray propagation while turning down native otel propagation formats. 

I would say that my proposal prefers maintaining the existing behavior in the majority of existing instrumentations so as to minimize the adverse impact to existing users while opening the door to support a wider category of users.  

> I think we should always prefer native otel propagation formats (like w3c traceparent) in lieu of user configuration.

I generally agree with this, though in this instance the principle of least surprise suggests that AWS Lambda tracing instrumentation should utilize tracing information provided by the AWS Lambda service.  I would agree with this in the case of Google Cloud Run as well.  More importantly, I think we need to be offering more user agency with regard to how instrumentation determines the carrier for context extraction when there can be uncertainty as to the desired behavior.

Reply:
> I think some of the flexibility across the semantic conventions is intentional in order to cover a broader range of (future) third-party instrumentations. I personally expect all instrumentations published under the OpenTelemetry org to follow all "Recommendeds" and SHOULDs, though I don't think we've had this discussion as a community (I've added this question to next week's semconv meeting).

I generally agree that [`SHOULD`](https://datatracker.ietf.org/doc/html/rfc2119#section-3) means "there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course."  I do not agree that there can be no other course taken as `SHOULD` is not [`MUST`](https://datatracker.ietf.org/doc/html/rfc2119#section-3).  I expect that maintaining compatibility while experimental specification language is being worked on to be a good reason for "choosing a different course".

Reply:
> I expect that maintaining compatibility while experimental specification language is being worked on to be a good reason for "choosing a different course".

👍

Reply:
(at the same time, adopting the latest semconv in instrumentations is the best way to get real user feedback which can be cycled back into semconv before it gets marked as stable)

Reply:
@jsuereth please see #277 for an attempt at language that would govern how instrumentation constructs a `Carrier` to be used with a user-specified `Propagator`.   I believe this should address all of your concerns expressed here.

Reply:
> (at the same time, adopting the latest semconv in instrumentations is the best way to get real user feedback which can be cycled back into semconv before it gets marked as stable)

I think that's what happened [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/7970).  The [real user feedback](https://github.com/open-telemetry/semantic-conventions/pull/277) was "this broke every existing usage of this instrumentation with no way to make it work again".